### PR TITLE
Sync EAGLE verify results across TP ranks

### LINKED ISSUE #28815: Speculative EAGLE verify: greedy/HIP branch missing TP broadcast -> rank divergence -> collective deadlock on ROCm
## Summary

On ROCm with TP>1 and speculative decoding (EAGLE/NEXTN), the EAGLE verify step can desynchronize TP ranks, producing different per-request accept counts across ranks → divergent running-batch composition → a subsequent collective (all-reduce) with mismatched shapes → **permanent TP deadlock** (all GPUs at 100% compute, 0% memory traffic; FastAPI still answers `/health` but `/v1/chat/completions` hangs).

This is the same failure *class* as #27221 / #27164 (TP rank-divergence into a collective), but a **different trigger**: not abort-on-timeout, but the speculative **verify** path.

## Root cause

In `python/sglang/srt/speculative/eagle_utils.py` (main @ `f42ec350b431`), the verify branch:

```python
if sampling_info.is_all_greedy or _is_npu or _is_hip:   # ROCm ALWAYS enters here
    target_predict = torch.argmax(next_token_logits, dim=-1)
    predict, accept_index, num_correct_drafts = verify_tree_greedy_func(...)
    # <-- NO cross-rank broadcast
else:
    ... tree_speculative_sampling_target_only(...)
    # the sampling branch DOES broadcast:
    if tp_group.world_size > 1:
        tp_group.broadcast(predict, src=0)
        tp_group.broadcast(accept_index, src=0)
        tp_group.broadcast(num_correct_drafts, src=0)
```

Because `_is_hip` forces the greedy branch on ROCm, the verify result (`predict`, `accept_index`, `num_correct_drafts`) is computed **locally per rank from `torch.argmax(next_token_logits)`** and is **never broadcast**. `next_token_logits` depends on the preceding all-reduce; when that all-reduce is **not bitwise-deterministic across ranks** (e.g. AMD aiter custom all-reduce, which is used on the speculative path even when the main path is on NCCL), ranks can pick different argmax on near-ties → different accept counts → divergent batch composition → deadlock on the next forward's collective.

The sampling branch already added a broadcast for exactly this reason ("different GPUs may produce slightly different ... causing different sampled tokens. Broadcast from rank 0 to ensure consistency."). The greedy branch needs the same.

## Proposed fix

Add the identical broadcast to the greedy branch after `verify_tree_greedy_func`:

```python
if sampling_info.is_all_greedy or _is_npu or _is_hip:
    target_predict = torch.argmax(next_token_logits, dim=-1)
    predict, accept_index, num_correct_drafts = verify_tree_greedy_func(...)
    tp_group = get_attention_tp_group() if is_dp_attention_enabled() else get_tp_group()
    if tp_group.world_size > 1:
        tp_group.broadcast(predict, src=0)
        tp_group.broadcast(accept_index, src=0)
        tp_group.broadcast(num_correct_drafts, src=0)
```

This guarantees rank-consistent accept results regardless of all-reduce determinism. Happy to open a PR.

## Environment

- SGLang `lmsysorg/sglang:v0.5.13.post1-rocm700-mi30x`; bug also present on `main` @ `f42ec350b431`.
- ROCm 7.0, 8x AMD MI300X (gfx942), TP=8, single node.
- Model: GLM-5.2-FP8 (`GlmMoeDsaForCausalLM`, DSA+MLA+MoE), `--speculative-algorithm NEXTN --speculative-num-steps 3 --speculative-num-draft-tokens 4`, `--kv-cache-dtype fp8_e4m3`.
- Startup logs: main forward `All-reduce call path: NCCL (custom AR disabled)`, but speculative path `Using AiterCustomAllreduce (AMD default)`.

## Reproduction

Concurrent, transition-heavy load (long decodes kept alive + long prefills joining + short request churn + streaming client-aborts + a few invalid payloads) wedges the server in ~5 min, even at low concurrency (1-2 running requests) — it is a batch-composition *transition* trigger, not raw overload. When wedged, all 8 TP scheduler processes are stuck at the same `run_batch` forward frame (py-spy), with GPUs at 100% use / 0% memory activity.

## Confirmed workaround

`--disable-custom-all-reduce` (forces deterministic NCCL all-reduce, so all ranks compute the same argmax). A/B on identical hardware/load: baseline wedges in ~5 min; with `--disable-custom-all-reduce` the same load ran 28 min with zero wedges. Setting `SYNC_TOKEN_IDS_ACROSS_TP=1` is a related backstop for the main-sampler path (`sampler.py`), though it does not cover the spec-verify accept counts.

### PR DESCRIPTION
**Pull request by zihaomu**
## Summary
- Move EAGLE verify result synchronization out of the sampling-only branch so greedy/HIP verify results are also broadcast across TP ranks.
- Keep `predict`, `accept_index`, and `num_correct_drafts` aligned before later running-batch updates and collectives.
- Add a CPU regression test that exercises the HIP-forced greedy path with a mocked TP group.

Fixes #28815.

## Motivation
On ROCm, EAGLE/NEXTN verify enters the greedy path via `_is_hip`. The sampling path already broadcasts verify outputs across TP ranks, but the greedy/HIP path returned local verify results. If ranks accept different numbers of draft tokens, later shape-dependent collectives can diverge and hang.

The fix treats TP synchronization as a verify-output invariant rather than a sampling-only behavior.

## Testing
- `python3 -m pytest test/registered/unit/spec/test_eagle_verify_tp_sync.py -q`
- `pre-commit run --files python/sglang/srt/speculative/eagle_utils.py test/registered/unit/spec/test_eagle_verify_tp_sync.py`
- MI300X/gfx942 ROCm container synthetic 2-rank validation: `python3 repro_issue_28815_tp_divergence.py --mode repo-fixed --master-port 29622 --join-timeout-seconds 30`

---
### CI States

Latest PR Test (Base): :x: [Run #28007911393](https://github.com/sgl-project/sglang/actions/runs/28007911393)
Latest PR Test (Extra): :x: [Run #28007911148](https://github.com/sgl-project/sglang/actions/runs/28007911148)

### CODE REVIEW

**Comment by ronhuafeng** [comment]
I did an independent CPU-side validation of this PR head (`f98b8edc56`).

Checks run:

```bash
git diff --check 067987d86^..067987d86
python3 -m py_compile python/sglang/srt/speculative/eagle_utils.py test/registered/unit/spec/test_eagle_verify_tp_sync.py
docker run --rm -v $PWD:/workspace -w /workspace lmsysorg/sglang:dev \
  bash -lc 'PYTHONPATH=python python3 -m pytest test/registered/unit/spec/test_eagle_verify_tp_sync.py -q'
```

Result: the focused test passed (`1 passed`). The patch moves the TP broadcast to the common verify-output path, so it covers the `_is_hip`/greedy branch described in #28815 instead of only the sampling branch. I did not run the full MI300X EAGLE/NEXTN workload here, but the added regression test matches the reported missing-broadcast invariant.

Also checked CI status: lint passed; the current red PR-test entries are gate/skip finish jobs rather than a focused lint/test failure.
